sql/spanner/inspect.go

```diff
@@ -0,0 +1,738 @@
+	i := strings.Index(x, "::"+r)
+	if i == -1 || !sqlx.IsQuoted(x[:i], '\'') {
+		return "", false
+	}
+	q := x[0:i]
+	x = x[1 : i-1]
+	switch t.Type.(type) {
+	case *schema.BoolType:
+		if sqlx.IsLiteralBool(x) {
+			return x, true
+		}
+	case *schema.DecimalType, *schema.IntegerType, *schema.FloatType:
+		if sqlx.IsLiteralNumber(x) {
+			return x, true
+		}
+	case *schema.BinaryType, *schema.JSONType, *schema.SpatialType, *schema.StringType, *schema.TimeType, *StringType, *BytesType:
+		return q, true
+	}
+	return "", false
+}
+
+func removeEmptyStrings(s []string) []string {
+	var r []string
+	for _, str := range s {
+		if str != "" {
+			r = append(r, str)
+		}
+	}
+	return r
+}
+
+type (
+
+	// columnDesc represents a column descriptor.
+	columnDesc struct {
+		typ       string
+		size      int
+		sizeIsMax bool
+	}
+
+	// SizeIsMax flags whether a column is of size "MAX" as opposed to a discreet int sizing.
+	SizeIsMax struct {
+		schema.Attr
+		Bool bool
+	}
+
+	// IsNullFiltered flags whether an index should be created
+	// with the NULL_FILTERED flag.
+	// https://cloud.google.com/spanner/docs/secondary-indexes#null-indexing-disable
+	IsNullFiltered struct {
+		schema.Attr
+		Bool bool
+	}
+
+	PartOrdinal struct {
+		schema.Attr
+		Int64 int64
+	}
+
+	// ArrayType defines an array type.
+	// https://www.spannerql.org/docs/current/arrays.html
+	ArrayType struct {
+		schema.Type
+		T string
+	}
+
+	// A BytesType represents a BYTES type.
+	BytesType struct {
+		schema.Type
+		T         string
+		Size      int
+		SizeIsMax bool
+	}
+
+	// CheckColumns attribute hold the column named used by the CHECK constraints.
+	// This attribute is added on inspection for internal usage and has no meaning
+	// on migration.
+	CheckColumns struct {
+		schema.Attr
+		Columns []string
+	}
+
+	// IndexType represents an index type.
+	// https://www.spannerql.org/docs/current/indexes-types.html
+	IndexType struct {
+		schema.Attr
+		T string // BTREE, BRIN, HASH, GiST, SP-GiST, GIN.
+	}
+
+	// IndexPredicate describes a partial index predicate.
+	// https://www.spannerql.org/docs/current/catalog-pg-index.html
+	IndexPredicate struct {
+		schema.Attr
+		P string
+	}
+
+	// IndexColumnProperty describes an index column property.
+	// https://www.spannerql.org/docs/current/functions-info.html#FUNCTIONS-INFO-INDEX-COLUMN-PROPS
+	IndexColumnProperty struct {
+		schema.Attr
+		// NullsFirst defaults to true for DESC indexes.
+		NullsFirst bool
+		// NullsLast defaults to true for ASC indexes.
+		NullsLast bool
+	}
+
+	// IndexStorageParams describes index storage parameters add with the WITH clause.
+	// https://www.spannerql.org/docs/current/sql-createindex.html#SQL-CREATEINDEX-STORAGE-PARAMETERS
+	IndexStorageParams struct {
+		schema.Attr
+		// AutoSummarize defines the authsummarize storage parameter.
+		AutoSummarize bool
+		// PagesPerRange defines pages_per_range storage
+		// parameter for BRIN indexes. Defaults to 128.
+		PagesPerRange int64
+	}
+
```

Comment:
To make this PR simpler to review and faster to merge, let's remove all types/parts that are not implemented in this this PR.